Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/branch.py

```diff
@@ -83,27 +87,67 @@ def editable(cls, branch, repository=None):
             return False
 
         remote_repo = repository.remote(name=repository.default_remote)
-        if remote_repo and isinstance(remote_repo, remote.GitHub):
-            for name in repository.source_remotes():
-                if branch in repository.branches_for(remote=name, cached=True):
-                    return False
+        if remote and remote_repo and isinstance(remote_repo, remote.GitHub):
+            return remote in repository.source_remotes()
         return True
 
     @classmethod
     def branch_point(cls, repository, limit=None):
         cnt = 0
-        commit = repository.commit(include_log=False, include_identifier=False)
-        while cls.editable(commit.branch, repository=repository):
-            cnt += 1
-            if limit and cnt > limit:
+        try:
+            proc = subprocess.Popen(
+                [repository.executable(), 'rev-list', 'HEAD'],
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+                cwd=repository.root_path,
+                **(dict(encoding='utf-8') if sys.version_info >= (3, 6) else dict())
+            )
+            if proc.poll():
                 return None
-            commit = repository.find(argument='HEAD~{}'.format(cnt), include_log=False, include_identifier=False)
-            if cnt > 1 or commit.branch != repository.branch or cls.editable(commit.branch, repository=repository):
-                log.info('    Found {}...'.format(string_utils.pluralize(cnt, 'commit')))
-            else:
-                log.info('    No commits on editable branch')
+            commit_hash = proc.stdout.readline().strip()
+            while commit_hash:
+                if limit and cnt > limit:
+                    return None
+
+                # If the specified commit is on the default branch, we've definitively found the branch point
+                if not run(
+                    [repository.executable(), 'merge-base', '--is-ancestor', commit_hash, repository.default_branch],
+                    cwd=repository.root_path,
+                    capture_output=True,
+                ).returncode:
+                    break
```

Comment:
I'm not particularly keen on calling this repeatedly in the hot loop? `merge-base` in any form is potentially expensive (as worst-case has to walk the entire parent DAG from `repository.default_branch` for each `commit_hash`).

I'd rather move `git merge-base HEAD {repository.default_branch}` to being outside the loop, and then we can simply check if we've reached the merge base yet within the loop.

We can also then limit the scope of the `rev-list` call to the commits between the merge base and HEAD.

Reply:
Oh, good point, yeah, we can avoid this call by doing that.